File: lib/python/openlinktoken-core-ai/src/test/openlinktoken/core/ai/tokentransformer/rotation/rotation_matrix_generator_test.py
# SPDX-License-Identifier: MIT

import threading

import numpy as np
import pytest

from openlinktoken.core.ai.tokentransformer.rotation.rotation_matrix_generator import generate

_IV = "test-rotation-iv-2024"
_ALT_IV = "different-iv-abc"
_DIMENSION = 4
_COUNT = 3


class TestRotationMatrixGenerator:
    """Unit tests for rotation matrix generation."""

    def test_returns_correct_count(self):
        """Generate exactly the requested number of matrices."""
        matrices = generate(_IV, _COUNT, _DIMENSION)
        assert len(matrices) == _COUNT

    def test_matrix_dimensions(self):
        """Ensure every generated matrix has the requested dimensions."""
        for matrix in generate(_IV, _COUNT, _DIMENSION):
            assert len(matrix) == _DIMENSION
            for row in matrix:
                assert len(row) == _DIMENSION

    def test_row_count_keeps_leading_rows_with_full_rotation_values(self):
        """A projected matrix should retain the leading rows of the full matrix."""
        full_matrix = generate(_IV, 1, _DIMENSION)[0]
        projected_matrix = generate(_IV, 1, _DIMENSION, row_count=2)[0]

        assert projected_matrix.shape == (2, _DIMENSION)
        np.testing.assert_array_equal(projected_matrix, full_matrix[:2, :])

    def test_row_count_must_be_within_dimension(self):
        """Invalid projected row counts should be rejected."""
        with pytest.raises(ValueError):
            generate(_IV, 1, _DIMENSION, row_count=0)
        with pytest.raises(ValueError):
            generate(_IV, 1, _DIMENSION, row_count=_DIMENSION + 1)

    def test_orthogonality(self):
        """Q @ Q^T must be the identity matrix within floating-point tolerance."""
        for matrix in generate(_IV, _COUNT, _DIMENSION):
            n = _DIMENSION
            for i in range(n):
                for j in range(n):
                    dot = sum(matrix[i][k] * matrix[j][k] for k in range(n))
                    expected = 1.0 if i == j else 0.0
                    assert abs(dot - expected) < 1e-10, f"Q @ Q^T[{i},{j}] = {dot}, expected {expected}"

    def test_proper_rotation_determinant(self):
        """det(Q) must equal +1 for a proper rotation matrix."""
        for matrix in generate(_IV, _COUNT, _DIMENSION):
            det = _det(matrix, _DIMENSION)
            assert abs(det - 1.0) < 1e-10, f"det(Q) = {det}, expected 1.0"

    def test_determinism_same_iv(self):
        """Same IV must always produce identical matrices."""
        matrices_a = generate(_IV, _COUNT, _DIMENSION)
        matrices_b = generate(_IV, _COUNT, _DIMENSION)
        for ma, mb in zip(matrices_a, matrices_b):
            assert np.array_equal(ma, mb)

    def test_different_ivs_produce_different_matrices(self):
        """Different IVs must produce different deterministic matrices."""
        matrices_a = generate(_IV, 1, _DIMENSION)
        matrices_b = generate(_ALT_IV, 1, _DIMENSION)
        assert not np.array_equal(matrices_a[0], matrices_b[0])

    def test_rotation_indices_differ(self):
        """Different rotation indices must produce different matrices."""
        matrices = generate(_IV, 3, _DIMENSION)
        assert not np.array_equal(matrices[0], matrices[1])
        assert not np.array_equal(matrices[1], matrices[2])

    def test_single_matrix(self):
        """Support generating a single rotation matrix."""
        matrices = generate(_IV, 1, _DIMENSION)
        assert len(matrices) == 1

    def test_rows_returns_only_leading_rows_without_changing_values(self):
        """A row limit should retain the exact leading rows of each full matrix."""
        full_matrices = generate(_IV, _COUNT, _DIMENSION)
        leading_rows = generate(_IV, _COUNT, _DIMENSION, rows=2)

        for actual, expected in zip(leading_rows, full_matrices):
            assert actual.shape == (2, _DIMENSION)
            np.testing.assert_array_equal(actual, expected[:2, :])

    def test_dimension_2(self):
        """Verify correctness for the smallest non-trivial dimension."""
        matrices = generate(_IV, 1, 2)
        matrix = matrices[0]
        n = 2
        for i in range(n):
            for j in range(n):
                dot = sum(matrix[i][k] * matrix[j][k] for k in range(n))
                expected = 1.0 if i == j else 0.0
                assert abs(dot - expected) < 1e-10
        det = _det(matrix, n)
        assert abs(det - 1.0) < 1e-10

    def test_dimension_8(self):
        """Preserve orthogonality and determinant for a larger dimension."""
        matrices = generate(_IV, 2, 8)
        n = 8
        for matrix in matrices:
            for i in range(n):
                for j in range(n):
                    dot = sum(matrix[i][k] * matrix[j][k] for k in range(n))
                    expected = 1.0 if i == j else 0.0
                    assert abs(dot - expected) < 1e-10
            det = _det(matrix, n)
            assert abs(det - 1.0) < 1e-10

    def test_matches_person_matching_qr_fixture(self):
        """Matrices must match the persisted PersonMatching QR-generation contract."""
        matrix = generate("qr-parity-fixture", 1, 8)[0]

        np.testing.assert_allclose(
            matrix[0],
            [
                -0.40747839357338234,
                -0.1408568566341346,
                -0.11969515964474949,
                0.2890408553442878,
                -0.6332317872229138,
                -0.03396738148325333,
                0.5422064923077496,
                0.14186260167654755,
            ],
            rtol=0.0,
            atol=1e-12,
        )

    def test_thread_safety(self):
        """Concurrent calls with the same IV must all produce the same result."""
        results = [None] * 100
        errors = []

        def worker(idx):
            """Generate one matrix concurrently and capture any exception."""
            try:
                results[idx] = generate(_IV, 1, _DIMENSION)[0]
            except Exception as e:
                errors.append(e)

        threads = [threading.Thread(target=worker, args=(i,)) for i in range(100)]
        for t in threads:
            t.start()
        for t in threads:
            t.join()

        assert not errors, f"Thread errors: {errors}"
        reference = results[0]
        for r in results[1:]:
            assert np.array_equal(r, reference)


def _det(matrix, n):
    """Compute determinant via Gaussian elimination for test validation."""
    a = [list(row) for row in matrix]
    sign = 1
    for col in range(n):
        max_row = col
        for row in range(col + 1, n):
            if abs(a[row][col]) > abs(a[max_row][col]):
                max_row = row
        if max_row != col:
            a[col], a[max_row] = a[max_row], a[col]
            sign = -sign
        for row in range(col + 1, n):
            factor = a[row][col] / a[col][col]
            for j in range(col, n):
                a[row][j] -= factor * a[col][j]
    result = sign
    for i in range(n):
        result *= a[i][i]
    return result

File: lib/java/openlinktoken-core-ai/src/test/java/org/openlinktoken/core/ai/tokentransformer/rotation/RotationMatrixGeneratorTest.java
/* SPDX-License-Identifier: MIT */
package org.openlinktoken.core.ai.tokentransformer.rotation;

import static org.junit.jupiter.api.Assertions.assertArrayEquals;
import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.junit.jupiter.api.Assertions.assertTrue;

import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.CountDownLatch;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

import org.junit.jupiter.api.Test;

/**
 * Tests deterministic orthogonal matrix generation and cross-language fixtures.
 */
class RotationMatrixGeneratorTest {

    private static final String IV = "test-rotation-iv-2024";
    private static final String ALT_IV = "different-iv-abc";
    private static final int DIMENSION = 4;
    private static final int COUNT = 3;
    private static final double TOLERANCE = 1e-10;
    private static final double QR_FIXTURE_TOLERANCE = 1e-15;

    @Test
    void testReturnsCorrectCount() {
        List<double[][]> matrices = RotationMatrixGenerator.generate(IV, COUNT, DIMENSION);
        assertEquals(COUNT, matrices.size());
    }

    @Test
    void testMatrixDimensions() {
        for (double[][] m : RotationMatrixGenerator.generate(IV, COUNT, DIMENSION)) {
            assertEquals(DIMENSION, m.length);
            for (double[] row : m) {
                assertEquals(DIMENSION, row.length);
            }
        }
    }

    @Test
    void testRowCountKeepsLeadingRowsWithFullRotationValues() {
        List<double[][]> fullMatrices = RotationMatrixGenerator.generate(IV, COUNT, DIMENSION);
        List<double[][]> projectedMatrices = RotationMatrixGenerator.generate(IV, COUNT, DIMENSION, 2);

        assertEquals(COUNT, projectedMatrices.size());
        for (int rotation = 0; rotation < COUNT; rotation++) {
            double[][] projected = projectedMatrices.get(rotation);
            double[][] full = fullMatrices.get(rotation);
            assertEquals(2, projected.length);
            for (int row = 0; row < projected.length; row++) {
                assertArrayEquals(full[row], projected[row]);
            }
        }
    }

    @Test
    void testRowCountEqualsDimensionMatchesFullApi() {
        List<double[][]> fullMatrices = RotationMatrixGenerator.generate(IV, COUNT, DIMENSION);
        List<double[][]> explicitFullMatrices =
                RotationMatrixGenerator.generate(IV, COUNT, DIMENSION, DIMENSION);

        for (int rotation = 0; rotation < COUNT; rotation++) {
            for (int row = 0; row < DIMENSION; row++) {
                assertArrayEquals(fullMatrices.get(rotation)[row], explicitFullMatrices.get(rotation)[row]);
            }
        }
    }

    @Test
    void testRowCountMustBeWithinDimension() {
        assertThrows(IllegalArgumentException.class,
                () -> RotationMatrixGenerator.generate(IV, COUNT, DIMENSION, 0));
        assertThrows(IllegalArgumentException.class,
                () -> RotationMatrixGenerator.generate(IV, COUNT, DIMENSION, DIMENSION + 1));
    }

    @Test
    void testOrthogonality() {
        for (double[][] m : RotationMatrixGenerator.generate(IV, COUNT, DIMENSION)) {
            int n = DIMENSION;
            for (int i = 0; i < n; i++) {
                for (int j = 0; j < n; j++) {
                    double dot = 0.0;
                    for (int k = 0; k < n; k++) {
                        dot += m[i][k] * m[j][k];
                    }
                    double expected = (i == j) ? 1.0 : 0.0;
                    assertTrue(
                            Math.abs(dot - expected) < TOLERANCE,
                            String.format("Q @ Q^T[%d,%d] = %f, expected %f", i, j, dot, expected));
                }
            }
        }
    }

    @Test
    void testProperRotationDeterminant() {
        for (double[][] m : RotationMatrixGenerator.generate(IV, COUNT, DIMENSION)) {
            double det = computeDet(m, DIMENSION);
            assertTrue(Math.abs(det - 1.0) < TOLERANCE,
                    String.format("det(Q) = %f, expected 1.0", det));
        }
    }

    @Test
    void testDeterminismSameIv() {
        List<double[][]> matricesA = RotationMatrixGenerator.generate(IV, COUNT, DIMENSION);
        List<double[][]> matricesB = RotationMatrixGenerator.generate(IV, COUNT, DIMENSION);
        for (int r = 0; r < COUNT; r++) {
            for (int row = 0; row < DIMENSION; row++) {
                for (int col = 0; col < DIMENSION; col++) {
                    assertEquals(matricesA.get(r)[row][col], matricesB.get(r)[row][col]);
                }
            }
        }
    }

    @Test
    void testDifferentIvsProduceDifferentMatrices() {
        double[][] a = RotationMatrixGenerator.generate(IV, 1, DIMENSION).get(0);
        double[][] b = RotationMatrixGenerator.generate(ALT_IV, 1, DIMENSION).get(0);
        boolean anyDiffers = false;
        outer:
        for (int row = 0; row < DIMENSION; row++) {
            for (int col = 0; col < DIMENSION; col++) {
                if (a[row][col] != b[row][col]) {
                    anyDiffers = true;
                    break outer;
                }
            }
        }
        assertTrue(anyDiffers, "Different IVs must produce different matrices");
    }

    @Test
    void testRotationIndicesDiffer() {
        List<double[][]> matrices = RotationMatrixGenerator.generate(IV, 3, DIMENSION);
        boolean zeroOneDiffers = false;
        boolean oneTwoDiffers = false;
        for (int row = 0; row < DIMENSION; row++) {
            for (int col = 0; col < DIMENSION; col++) {
                if (matrices.get(0)[row][col] != matrices.get(1)[row][col]) {
                    zeroOneDiffers = true;
                }
                if (matrices.get(1)[row][col] != matrices.get(2)[row][col]) {
                    oneTwoDiffers = true;
                }
            }
        }
        assertTrue(zeroOneDiffers, "Matrix 0 and matrix 1 must differ");
        assertTrue(oneTwoDiffers, "Matrix 1 and matrix 2 must differ");
    }

    @Test
    void testDimension2() {
        List<double[][]> matrices = RotationMatrixGenerator.generate(IV, 1, 2);
        double[][] m = matrices.get(0);
        int n = 2;
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                double dot = 0.0;
                for (int k = 0; k < n; k++) {
                    dot += m[i][k] * m[j][k];
                }
                double expected = (i == j) ? 1.0 : 0.0;
                assertTrue(Math.abs(dot - expected) < TOLERANCE);
            }
        }
        assertTrue(Math.abs(computeDet(m, n) - 1.0) < TOLERANCE);
    }

    @Test
    void testDimension8() {
        int n = 8;
        for (double[][] m : RotationMatrixGenerator.generate(IV, 2, n)) {
            for (int i = 0; i < n; i++) {
                for (int j = 0; j < n; j++) {
                    double dot = 0.0;
                    for (int k = 0; k < n; k++) {
                        dot += m[i][k] * m[j][k];
                    }
                    double expected = (i == j) ? 1.0 : 0.0;
                    assertTrue(Math.abs(dot - expected) < TOLERANCE);
                }
            }
            assertTrue(Math.abs(computeDet(m, n) - 1.0) < TOLERANCE);
        }
    }

    @Test
    void testMatchesNumpyHouseholderQrReferenceFixture() {
        // NumPy QR output after normalizing columns by sign(diag(R)) and enforcing det(Q) = +1.
        double[][] expected = {
                { -0.11033054293072242, -0.65094870327709264, 0.73763853752775732, 0.14135892243645001 },
                { 0.45120119662062919, -0.059568165581975238, 0.18195883986509798, -0.87164218255672898 },
                { -0.86485909149250417, -0.11280580567809925, -0.13902921534993631, -0.46900370931076435 },
                { -0.19042952325593890, 0.74832631221705959, 0.63517812133923512, -0.017119617054466726 }
        };

        double[][] actual = RotationMatrixGenerator.generate(IV, 1, DIMENSION).get(0);

        for (int row = 0; row < DIMENSION; row++) {
            for (int col = 0; col < DIMENSION; col++) {
                assertEquals(expected[row][col], actual[row][col], QR_FIXTURE_TOLERANCE,
                        "Q[" + row + "][" + col + "]");
            }
        }
    }

    @Test
    void testThreadSafety() throws InterruptedException {
        int threadCount = 100;
        List<double[][]> results = new ArrayList<>(threadCount);
        for (int i = 0; i < threadCount; i++) {
            results.add(null);
        }
        List<Exception> errors = new ArrayList<>();
        CountDownLatch latch = new CountDownLatch(threadCount);

        ExecutorService pool = Executors.newFixedThreadPool(threadCount);
        for (int i = 0; i < threadCount; i++) {
            final int idx = i;
            pool.submit(() -> {
                try {
                    results.set(idx, RotationMatrixGenerator.generate(IV, 1, DIMENSION).get(0));
                } catch (Exception e) {
                    synchronized (errors) {
                        errors.add(e);
                    }
                } finally {
                    latch.countDown();
                }
            });
        }
        latch.await();
        pool.shutdown();

        assertTrue(errors.isEmpty(), "Thread errors: " + errors);
        double[][] reference = results.get(0);
        for (double[][] r : results) {
            for (int row = 0; row < DIMENSION; row++) {
                for (int col = 0; col < DIMENSION; col++) {
                    assertEquals(reference[row][col], r[row][col]);
                }
            }
        }
    }

    /** Compute determinant via Gaussian elimination for test validation. */
    private static double computeDet(double[][] matrix, int n) {
        double[][] a = new double[n][];
        for (int i = 0; i < n; i++) {
            a[i] = matrix[i].clone();
        }
        double sign = 1.0;
        for (int col = 0; col < n; col++) {
            int maxRow = col;
            for (int row = col + 1; row < n; row++) {
                if (Math.abs(a[row][col]) > Math.abs(a[maxRow][col])) {
                    maxRow = row;
                }
            }
            if (maxRow != col) {
                double[] tmp = a[col];
                a[col] = a[maxRow];
                a[maxRow] = tmp;
                sign = -sign;
            }
            for (int row = col + 1; row < n; row++) {
                double factor = a[row][col] / a[col][col];
                for (int j = col; j < n; j++) {
                    a[row][j] -= factor * a[col][j];
                }
            }
        }
        double det = sign;
        for (int i = 0; i < n; i++) {
            det *= a[i][i];
        }
        return det;
    }
}
